internal/client/job.go

```diff
@@ -235,21 +236,63 @@ func (c *Project) doJobMonitored(ctx context.Context, job *pb.Job, ui terminal.U
 		}
 	} else {
 		var configRunner *configpkg.Runner
-		// Find runner configuration on the app
+		// Find runner config on the app
 		if job.Application != nil {
 			configRunner = c.waypointHCL.ConfigAppRunner(job.Application.Application)
 		}
-		// If not on app, try to find it on the project
-		if configRunner == nil {
+		// Find runner config on the project
+		if configRunner != nil {
+			if c.waypointHCL.ConfigRunner() != nil {
+				log.Warn("runner configurations found on both app and project, app takes precedence.")
+			}
+		} else {
 			configRunner = c.waypointHCL.ConfigRunner()
 		}
-		// If runner config is found, assign to job
+		// If runner config exists, assign to job
 		if configRunner != nil {
 			if configRunner.Profile != "" {
 				job.OndemandRunner = &pb.Ref_OnDemandRunnerConfig{
 					Name: configRunner.Profile,
 				}
 			}
+			if configRunner.TargetLabels != nil {
+				if job.OndemandRunner != nil {
+					log.Warn("both runner profile and target labels are set, profile takes precedence.")
+				} else {
+					runners, err := c.client.ListRunners(ctx, &pb.ListRunnersRequest{})
+					if err != nil {
+						return nil, err
+					}
+					if len(runners.Runners) == 0 {
+						return nil, status.Errorf(codes.NotFound, "no runners found.")
+					}
+					targeted := false
+					for _, r := range runners.Runners {
+						n := len(configRunner.TargetLabels)
+						// Check whether a runner matches every target label
+						// If multiple runners match labels, this will take the first that matches
+						for k, v := range configRunner.TargetLabels {
+							if val, ok := r.Labels[k]; ok && v == val {
+								n -= 1
+							}
+						}
```

Comment:
@xx and I paired on this - We can simplify this check for looking at when `TargetLabels` has to be a subset of `Runners` with a bool